### PR TITLE
Expand warnings output for ResourceWarning

### LINKED ISSUE #9644: Expand warnings output for ResourceWarning
It is difficult to debug ``ResourceWarning`` from the existing output.

In [Python3.6 a ``source`` field was added](https://bugs.python.org/issue26567) to the ``warnings.WarningMessage`` objects.

Since pytest now requires Python 3.7 I propose the following addition to the existing ``warning_record_to_str``:

```python
def warning_record_to_str(warning_message: warnings.WarningMessage) -> str:
    """Convert a warnings.WarningMessage to a string."""
    warn_msg = warning_message.message
    msg = warnings.formatwarning(
        str(warn_msg),
        warning_message.category,
        warning_message.filename,
        warning_message.lineno,
        warning_message.line,
    )
    tb = tracemalloc.get_object_traceback(warning_message.source)
    if tb is not None:
        msg = "%sSource of unclosed object:\n%s" % (msg, '\n'.join(tb.format()))
    elif issubclass(warning_message.category, ResourceWarning):
        # tracemalloc is inactive
        # TODO needs to be adjusted, see below in ticket
        msg = "%sGet a traceback for where the unclosed object was allocated by ..." % msg
    return msg
```

The output looks like this as an example:
```
  /.../devpi/devel/lib/python3.8/enum.py:315: ResourceWarning: unclosed <socket.socket [closed] fd=53, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6>
    return cls.__new__(cls, value)
  Source of unclosed object:
    File "/.../devpi/devel/lib/python3.8/site-packages/urllib3/connection.py", line 205
      conn = self._new_conn()
    File "/.../devpi/devel/lib/python3.8/site-packages/urllib3/connection.py", line 174
      conn = connection.create_connection(
    File "/.../devpi/devel/lib/python3.8/site-packages/urllib3/util/connection.py", line 77
      sock = socket.socket(af, socktype, proto)
```

This requires that ``tracemalloc`` is active. So the open question is, when should ``tracemalloc.start()`` be called and how to pass the size of the stack trace to it, as the default of ``1`` is often too small (I needed 20 to find the source of an unclosed requests Response). I propose three possible solutions:

1. Document running pytest like this: ``python -X tracemalloc=20 -m pytest ...``
2. Add a command line option to enable ``tracemalloc`` with the limit as parameter.
3. Enable tracing when warnings for ``ResourceWarning`` are set to ``once`` or ``always`` and always use a stack trace limit of ``3``.

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #9644

cc @fschulze

### CODE REVIEW

File: src/_pytest/warnings.py [hunk 1]
```diff
@@ -81,6 +81,25 @@ def warning_record_to_str(warning_message: warnings.WarningMessage) -> str:
...
+        except ImportError:
+            pass
+        else:
+            tb = tracemalloc.get_object_traceback(warning_message.source)
+            if tb is not None:
+                formatted_tb = "\n".join(tb.format())
+                # Use a leading new line to better separate the (large) output
+                # from the traceback to the previous warning text.
+                msg += (
+                    f"\nObject allocated at (most recent call first):\n{formatted_tb}"
```
Comment (fschulze) [inline src/_pytest/warnings.py:96]: I'm not sure ``most recent call first`` is correct here, maybe just remove it?

File: src/_pytest/warnings.py [hunk 2]
```diff
@@ -81,6 +81,25 @@ def warning_record_to_str(warning_message: warnings.WarningMessage) -> str:
         warning_message.lineno,
         warning_message.line,
     )
+    if warning_message.source is not None:
+        try:
+            import tracemalloc
```
Comment (fschulze) [inline src/_pytest/warnings.py:86]: Are there platforms which don't have ``tracemalloc`` available?

**Review by fschulze — APPROVED** [review]
Nice! Thanks for the PR, you beat me to it :wink:

File: src/_pytest/warnings.py (hunk 2 above)
Comment (fschulze) [reply src/_pytest/warnings.py:86]: It is likely that PyPy doesn't support it at all due to differing memory management.

File: testing/test_warnings.py [hunk 3]
```diff
@@ -774,3 +775,53 @@ def test_it():
...
+    # available, using `importorskip("tracemalloc")` for example,
+    # because we want to ensure the same code path does not break in those platforms.
+    try:
+        import tracemalloc  # noqa
+
+        has_tracemalloc = True
+    except ImportError:
+        has_tracemalloc = False
+
+    pytester.makepyfile(
```
Comment (bluetech) [inline testing/test_warnings.py:792]: I think this test (tracemalloc not enabled) will fail if the person running the *pytest* test suite enables tracemalloc.

**Review by bluetech — APPROVED** [review]
Looks great to me! I left one comment.

File: testing/test_warnings.py (hunk 3 above)
Comment (nicoddemus) [reply testing/test_warnings.py:792]: Hmm why do you say that? And do you mean running in CPython?

I tried it here:

```
python -X tracemalloc=20 -m pytest testing\test_warnings.py
```

And it passes.

Comment (nicoddemus) [reply testing/test_warnings.py:792]: Note the try/except is checking if tracemalloc is available at all, which is not on PyPy for example, not if tracemalloc is enabled or not.

Comment (bluetech) [reply testing/test_warnings.py:792]: I guess `python -X tracemalloc=20` doesn't transfer, but an envvar does, so try this:

```
PYTHONTRACEMALLOC=100 pytest testing/test_warnings.py -k resource
```

Comment (nicoddemus) [reply testing/test_warnings.py:792]: Ahh good catch. 😁 

Updated the test, thanks.

**Pull request merged**